packages/concerto-tools/lib/codegen/fromcto/odata/odatavisitor.js

```diff
@@ -0,0 +1,317 @@
+                parameters.fileWriter.writeLine(1, `<edmx:Include Namespace="${clazz.getNamespace()}" />`);
+                parameters.fileWriter.writeLine(0, '</edmx:Reference>');
+            }
+        }
+
+        parameters.fileWriter.writeLine(0, '<edmx:DataServices>');
+        parameters.fileWriter.writeLine(1, `<Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="${modelFile.getNamespace()}">`);
+
+        modelFile.getAllDeclarations().forEach((decl) => {
+            decl.accept(this, parameters);
+        });
+
+        parameters.fileWriter.writeLine(1, `<EntityContainer Name="${modelFile.getNamespace()}Service">`);
+        modelFile.getAllDeclarations().forEach((decl) => {
+            if (!decl.isAbstract() && decl.isIdentified()) {
+                parameters.fileWriter.writeLine(2, `<EntitySet Name="${decl.getName()}" EntityType="${decl.getFullyQualifiedName()}"/>`);
+            }
+        });
+        parameters.fileWriter.writeLine(1, '</EntityContainer>');
+
+        parameters.fileWriter.writeLine(1, '</Schema>');
+        parameters.fileWriter.writeLine(0, '</edmx:DataServices>');
+        parameters.fileWriter.writeLine(0, '</edmx:Edmx>');
+        parameters.fileWriter.closeFile();
+
+        return null;
+    }
+
+    /**
+     * Visitor design pattern
+     * @param {ClassDeclaration} classDeclaration - the object being visited
+     * @param {Object} parameters  - the parameter
+     * @return {Object} the result of visiting or null
+     * @private
+     */
+    visitEnumDeclaration(classDeclaration, parameters) {
+        parameters.fileWriter.writeLine(2, `<EnumType Name="${classDeclaration.getName()}">`);
+        classDeclaration.getOwnProperties().forEach((property) => {
+            property.accept(this, parameters);
+        });
+        parameters.fileWriter.writeLine(2, '</EnumType>');
+        return null;
+    }
+
+    /**
+     * Visitor design pattern
+     * @param {ClassDeclaration} classDeclaration - the object being visited
+     * @param {Object} parameters  - the parameter
+     * @return {Object} the result of visiting or null
+     * @private
+     */
+    visitClassDeclaration(classDeclaration, parameters) {
+        const stereoType = classDeclaration.isIdentified() ? 'EntityType' : 'ComplexType';
+        const abstract = classDeclaration.isAbstract() ? 'Abstract="true"' : '';
+        let superType = '';
+
+        if (classDeclaration.getSuperType()) {
+            superType = `BaseType="${classDeclaration.getSuperType()}"`;
+        }
+        parameters.fileWriter.writeLine(2, `<${stereoType} Name="${classDeclaration.getName()}" ${abstract} ${superType}>`);
+
+        classDeclaration.getDecorators().forEach(decorator => {
+            decorator.accept(this, parameters);
+        });
+
+        /**
+         * Handle Keys (identifiers).
+         * In OData a type can only specify a Key if not already specified by a super-type.
```

Comment:
Don't we have the same constraint in Concerto?

Reply:
Yes